core/src/test/scala/unit/kafka/log/LogTest.scala

```diff
@@ -1013,6 +1013,54 @@ class LogTest extends JUnitSuite {
   }
 
   @Test
+  def testOverCompactedLogRecovery(): Unit = {
+    // append some messages to create some segments
+    val logProps = new Properties()
+    logProps.put(LogConfig.SegmentBytesProp, 1000: java.lang.Integer)
+    logProps.put(LogConfig.MaxMessageBytesProp, 64*1024: java.lang.Integer)
+    logProps.put(LogConfig.IndexIntervalBytesProp, 1: java.lang.Integer)
+    val config = LogConfig(logProps)
+    val log = new Log(logDir,
+      config,
+      recoveryPoint = 0L,
+      time.scheduler,
+      time)
+    val set1 = new ByteBufferMessageSet(
+      compressionCodec = NoCompressionCodec,
+      offsetSeq = Seq(0),
+      messages = new Message("v1".getBytes(), "k1".getBytes(), Message.NoTimestamp, magicValue = Message.CurrentMagicValue))
+    val set2 = new ByteBufferMessageSet(
```

Comment:
@michaelschiff Currently the offset within the same MessageSet would not differ by more than Integer.MAX_VALUE. Otherwise the message would already hit the max message size. During log compaction we do not merge MessageSet either. So maybe we don't need to worry too much about this.

If we assume the above is the true. As a short term fix, when log roll happens, we can always roll a new log segment with base offset `max(logEndOffset, largestOffsetOfTheMessageSetToAppend - Integer.MAX_VALUE)`. So in the case of what @junrao raised: two consecutive message offsets of 0 and `Integer.MAX_VALUE.toLong + 2`, the newly rolled out segment will have `max(1, Integer.MAX_VALUE.toLong + 2 - Integer.MAX_VALUE.toLong)` which is 2.

This is a little hacky but seems working.